---
title: "Lineage Reconstruction"
author: "Kelly Street"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteEncoding{UTF-8}
---

<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{slingshot Vignette}
-->

```{r options, results="hide", include=FALSE, cache=FALSE, results='hide', message=FALSE}
## change cache to FALSE
knitr::opts_chunk$set(fig.align="center", cache=TRUE, cache.path = "slingshotTutorial_cache/", fig.path="slingshotTutorial_figure/",error=FALSE, #make it stop on error
fig.width=6,fig.height=6,autodep=TRUE,out.width="600px",out.height="600px", results="markup", echo=TRUE, eval=TRUE)
#knitr::opts_knit$set(stop_on_error = 2L) #really make it stop
#knitr::dep_auto()
options(getClass.msg=FALSE) #get rid of annoying messages about cache until fixed internally in R

set.seed(6473) ## for reproducibility

## library(bioc2016singlecell) ## add back when ready

## for now load individual dependencies
library(slingshot)
```

# Introduction

This is the third part of the Bioc2016 workshop "Analysis of single-cell RNA-seq data with R and Bioconductor."

In this part we will cover lineage reconstruction with the `Githubpkg("kstreet13/slingshot")` package.

The goal of `slingshot` is to use clusters to uncover global structure and convert this structure into smooth lineages represented by one-dimensional variables, often called ``pseudotime.'' We give tools for learning cluster relationships in an unsupervised or semi-supervised manner and constructing smooth curves representing each lineage, with visualization methods for each step.

## Basic `slingshot` analysis

The minimal input to `slingshot` is a matrix representing the cells in a reduced-dimensional space and a vector of clustering results. The analysis then procedes:

* Find connections between clusters with the `get_lineages` function, optionally specifying known start and end points.
* Construct smooth curves and pseudotime variables with the `get_curves` function.
* Assess the cluster connectivity with `plot_tree` and the curve stability with `plot_curves`.

Using connections between clusters to define global structure improves the stability of inferred lineages. And our use of smooth curves in place of piecewise-linear ones reduces variability in the inferred pseudotime vectors.

## An example dataset

We will take our inputs from the previous sections: the normalized counts matrix obtained with `scone` and the cluster assignments obtained with `clusterExperiment`, both of which can be loaded directly from the workshop package.

```{r datain, eval=TRUE}
## data(normClust) ...eventually
# clus <- clusterMatrix(ce)[,1]  is this right?

## for now
load('~/Projects/oe_p63/E4_scone_none_fq_ruv1_bio_nobatch_1Kgl05.Rda')
X <- pcaX[! mergeCl %in% c(4,20), 1:3]
clus <- mergeCl[! mergeCl %in% c(4,20)]

## Dimensionality reduction
# pcaX <- prcomp(t(counts), center = TRUE, scale = FALSE)
```

# Step 1: Assign clusters to lineages with `get_lineages`

The `get_lineages` function takes as input an `n x p` matrix and a vector of clustering results of length `n`. It then maps connections between adjacent clusters using a minimum spanning tree (MST) and identifies paths through these connections that represent potential lineages.

This analysis can be performed in an entirely unsupervised manner or in a semi-supervised manner by specifying known beginning and end point clusters. We recommend that you specify a root cluster; this will have no effect on how the clusters are connected, but it will allow for nicer curves in datasets with a branching structure. Pre-specified end point clusters will be constrained to only one connection.

```{r unsup_lines}
l1 <- get_lineages(X, clus)
# plot_tree(X, clus, l1, threeD = TRUE)
plot_tree(X, clus, l1, dim = 3)
```

Running `get_lineages` with no supervision produces the connections shown above. Since no root cluster was specified, `slingshot` picked one of the leaf-node clusters to be the beginning, based on a simple parsimony rule. The root cluster is the leaf-node cluster connected by a green line.

```{r sup_lines_start}
l2 <- get_lineages(X, clus, start.clus = '10')
# plot_tree(X, clus, l2, threeD = TRUE)
plot_tree(X, clus, l2, dim = 3)
```

When we specify a root cluster we get the same connections and the only difference is which line is drawn in green.

```{r sup_lines_end}
l3 <- get_lineages(X, clus, start.clus = '10', end.clus = '17')
# plot_tree(X, clus, l3, threeD = TRUE)
plot_tree(X, clus, l3, dim = 3)
```

Here we demonstrate the ability to specify end point clusters, which puts a constraing on the connections. We now draw the MST subject to the constraint that given end point clusters must be leaves. Pre-specified end point clusters are connected by red lines.

There are a few additional arguments we could have passed to `get_lineages` for more greater control:
* `dist.fun` is a function for computing distances between clusters. The default is squared distance between cluster centers normalized by their joint covariance matrix.
* `omega` is a granularity parameter, allowing the user to set an upper limit on connection distances. It takes values between 0 and 1 (or `Inf`), representing a percentage of the largest observed distance.
* `distout` is a logical value, indicating whether the user wants the pairwise cluster distance matrix to be returned with the output.

After constructing the MST, `get_lineages` identifies paths through the tree to designate as lineages. At this stage, a lineage will consist of an ordered set of cluster names, starting with the root cluster and ending with a leaf. The output of `get_lineages` is a list of these vectors, along with some additional information on how they were constructed.

# Step 2: Construct smooth lineages and order cells with `get_curves`

To find a consensus clustering across the many different clusterings created by `clusterMany` the function `combineMany` can be used next. 

```{r combineMany, eval=FALSE}
ce <- combineMany(ce, proportion = 0.6, minSize = 5)
```

Notice we get a warning that we did not specify any clusters to combine, so it is using the default -- those from the previous call. 

If we look at the `clusterMatrix` of the returned `ce` object, we see that the new cluster from `combineMany` has been added to the existing clusterings. This is the basic strategy of the functions in this package. Any clustering that is created is added to existing clusterings, so the user does not need to keep track of past clusterings and can easily compare what has changed. 

```{r lookAtCombineMany, eval=FALSE}
head(clusterMatrix(ce)[,1:3])
par(mar=plotCMar)
plotClusters(ce)
```

The proportion argument regulates how many times two samples need to be in the same cluster across parameters to be together in the combined clustering. Decreasing the value of `proportion` results in fewer "unclustered" (i.e., -1) samples. Another parameter that controls the number of unassigned samples is `minSize`, which discards the combined clusters with less than `minSize` samples.

# Step 3: Find temporally expressed genes

It is not uncommon that `combineMany` will result in too many small clusters, which in practice are too closely related to be useful. Since our final goal is to find gene markers for each clusters, we argue that we can merge clusters that show no or little differential expression (DE) between them.

This functionality is implemented in the `mergeClusters` function. `mergeClusters` needs a hierarchical clustering of the clusters; it then goes progressively up that hierarchy, deciding whether two adjacent clusters can be merged. The function `makeDendrogram` makes such a hierarchy between clusters (by applying `hclust` to the medoids of the clusters).

Here, we use the 1,000 most variable genes to make the cluster hierarchy.

```{r makeDendrogram, eval=FALSE}
ce<-makeDendrogram(ce, dimReduce="var", ndims=1000)
plotDendrogram(ce)
```

It is useful to first run `mergeClusters` without actually creating any object so as to preview what the final clustering will be (and perhaps to help in setting the cutoff).

```{r mergeClustersPlot, eval=FALSE}
mergeClusters(ce, mergeMethod="adjP", plot="mergeMethod")
```

```{r mergeClusters, eval=FALSE}
ce <- mergeClusters(ce,mergeMethod="adjP",cutoff=0.01)
par(mar=plotCMar)
plotClusters(ce)
plotCoClustering(ce,whichClusters=c("mergeClusters","combineMany"))
```

Notice that `mergeClusters` combines clusters based on the actual values of the features, while the `coClustering` plot shows how often the samples clustered together.

Finally, we can do a heatmap visualizing this final step of clustering.

```{r plotHeatmap, eval=FALSE}
plotHeatmap(ce, clusterSamplesData="dendrogramValue", breaks=.99)
```


# Session Info

```{r session}
sessionInfo()
```
